design/baremetal-operator/maintenance-mode.md

```diff
@@ -0,0 +1,87 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Add support for Maintenance Mode flag
+
+## Status
+
+implementable
+
+## Summary
+
+Currently there is no way to prevent management of BaremetalHost resources
+after provisioning is completed.  The underlying Ironic API contains a
+maintenance-mode flag which could be used to disable management of
+BMH resources.
+
+## Motivation
+
+In a multi-tier deployment where one cluster deploys another, the "parent"
+cluster will contain BMH resources for initial provisioning,
+but the "child" cluster will later contain BMH resources that reference the
+same physical hosts.
+
+In this scenario it's necessary to prevent management operations from the
+parent cluster such as asserting power state, or BMH actions on the child
+cluster such as the [reboot annotation](reboot-interface.md) may fail due
+to unwanted BMC interactions from the parent cluster.
+
+This may also be useful for temporary maintenance operations
+where you may wish for power management operations and polling
+to be stopped for specific hosts (this is the intended use if the
+underlying [Ironic API](https://docs.openstack.org/api-ref/baremetal/?expanded=set-maintenance-flag-detail#set-maintenance-flag)
```

Comment:
It is, as is written here, the intended use case of maintenance, but only if we're ready to avoid touching ironic API while in maintenance. E.g. trying to do adoption in maintenance mode is a huge stretch already, I'm not even sure it works (even less sure if it should work).

The alternative suggestion: removes ironic nodes (not BMHs!) that the current cluster should not be managing. This way there is only one ironic responsible for each node, and we don't have any split-brain issues.

WDYT?

Reply:
I'm primarily trying to solve the "cluster deploys a cluster" problem, where we need to pivot ownership of the BMH from the parent->child cluster (but maintain the BMH in the parent for inventory management and so redeployment could be possible)

In that case, some way to remove the nodes from Ironic on the parent makes sense, but I'm not sure the `spec.maintenance: true` API really describes this behavior.

Maybe we should add an annotation e.g `baremetalhost.metal3.io/unmanage` which triggers removal from Ironic, and makes all subsequent management operations a no-op?

@kashifest @maelk @dhellmann what are your thoughts?

Reply:
> It is, as is written here, the intended use case of maintenance, but only if we're ready to avoid touching ironic API while in maintenance. E.g. trying to do adoption in maintenance mode is a huge stretch already, I'm not even sure it works (even less sure if it should work).

What sequence of events would lead to trying to adopt a host with the maintenance flag on?

Should we treat hosts with their maintenance flag set as paused, except to allow delete?

> The alternative suggestion: removes ironic nodes (not BMHs!) that the current cluster should not be managing. This way there is only one ironic responsible for each node, and we don't have any split-brain issues.

Are you suggesting that the maintenance flag should cause the BMO to remove the host from Ironic completely? Or that we should describe another way to do that instead of (or in addition to) adding a maintenance flag?

> Maybe we should add an annotation e.g `baremetalhost.metal3.io/unmanage` which triggers removal from Ironic, and makes all subsequent management operations a no-op?

I think that's more or less what the existing pause annotation was meant to be, but we didn't think through the fact that by not removing the host from Ironic we leave it's periodic tasks operating. So we could extend what the existing annotation does, too, if we decide we need to go this route.

Does anyone else run into this problem of wanting to deploy repeatedly from a persistent cluster? Or do most users run the deployment from a minikube or other disposable cluster?

Reply:
Is there some reason we'd want the systems to remain in Ironic? In case we kept them in BMO but removed them from Ironic with `baremetalhost.metal3.io/unmanage`, what would happen when the target cluster is done with the node and we remove the `baremetalhost.metal3.io/unmanage` and we add the system to Ironic again? Would that trigger BMC actions?

Reply:
> What sequence of events would lead to trying to adopt a host with the maintenance flag on?

That will happen any time the metal3 pod gets rescheduled, assuming we leave the hosts in Ironic.
 
> Should we treat hosts with their maintenance flag set as paused, except to allow delete?

We could, I was thinking it might be useful to enable some actions, in particular to update the hardwaredetails via the [new annotation](https://github.com/metal3-io/baremetal-operator/blob/master/docs/inspectAnnotation.md) - so that the details of each node are visible on the parent cluster.

Pausing would also been reasonable I guess, it'd just mean e.g UX on the parent cluster would have to interact with the child-cluster BMH (which is probably necessary for other reasons).

Pausing is also probably easier to implement than skipping all states except delete?

> Are you suggesting that the maintenance flag should cause the BMO to remove the host from Ironic completely? Or that we should describe another way to do that instead of (or in addition to) adding a maintenance flag?

I'm suggesting we could describe another API that does that (such as an "unmanage" annotation), instead of the maintenance flag (although that could also be added if there's a requirement for it)

Reply:
> Is there some reason we'd want the systems to remain in Ironic? 

I don't think so, the main reason I proposed the maintenance flag is it looks like the simplest option to implement.

> In case we kept them in BMO but removed them from Ironic with `baremetalhost.metal3.io/unmanage`, what would happen when the target cluster is done with the node and we remove the `baremetalhost.metal3.io/unmanage` and we add the system to Ironic again? Would that trigger BMC actions?

If the parent cluster removes the annotation (or any other flag we add here), I would expect the host to be added to Ironic again, and that will trigger BMC actions.

This may be a necessary option in the case where the parent cluster has to redeploy a child cluster, but in all other cases probably won't happen, since the child cluster BMH resources are managing the hardware?

Reply:
> > What sequence of events would lead to trying to adopt a host with the maintenance flag on?
> 
> That will happen any time the metal3 pod gets rescheduled, assuming we leave the hosts in Ironic?

Sure. But in that case, Ironic doesn't think the host is in maintenance mode and we can avoid putting it into maintenance mode until we get to a steady state. I think this point is probably moot, though, if we're going to reuse the pause annotation.

> > Are you suggesting that the maintenance flag should cause the BMO to remove the host from Ironic completely? Or that we should describe another way to do that instead of (or in addition to) adding a maintenance flag?
> 
> I'm suggesting we could describe another API that does that (such as an "unmanage" annotation), instead of the maintenance flag (although that could also be added if there's a requirement for it)

I still view Ironic as an implementation detail of metal3. It's one that would be extremely difficult to change, but I don't think we want to treat metal3 as "Ironic for k8s". Exposing the state of whether we have synced a host into Ironic's database leaks all sorts of things through our API that are best kept hidden so we can change them -- the fact that Ironic has a separate view of the state of the world, the fact that we deploy Ironic without stable storage, the fact that we have to restore Ironic's knowledge of a host before we can do virtually anything with it, etc.

Reply:
> Sure. But in that case, Ironic doesn't think the host is in maintenance mode and we can avoid putting it into maintenance mode until we get to a steady state. I think this point is probably moot, though, if we're going to reuse the pause annotation.

I was thinking we'd have to support reschedule (and thus adopt) of hosts that are in maintenance mode, if we went with that option, but as you say it seems like reworking the paused behavior is the preferred option which conveniently avoids dealing with that problem ;)